### PR TITLE
Add a playbook to ensure the base image exists

### PR DESCRIPTION
**Pull request by smarterclayton**
Experimentation with a base image building playbook that can be included
from the launch playbook or run independently.

### CODE REVIEW

**Comment by smarterclayton** [comment]
@sdodson @abutcher @stevekuznetsov this is a very rough cut, but here's the structure i'm imagining for provisioning images for CI and gold images - if there is existing work for this please link it:

1. playbook that builds the image from a base image, a set of roles, and then always writes the image
   1. performs registration and yum updates if necessary
   2. roles for the items under this
2. top level play in `launch.yaml` that checks for image existence and includes the playbook if necessary to provision the image the first time

I don't see a way to structure image provisioning differently - we need to create a disk, launch an instance, and then want to run roles on that instance.  That requires a top level playbook (since we're moving from local to remote).

**Comment by stevekuznetsov** [comment]
Origin CI Tool instance preparation logic is [here](https://github.com/openshift/origin-ci-tool/blob/master/oct/ansible/oct/playbooks/prepare/main.yml).

**Comment by stevekuznetsov** [comment]
Performance testing team image minting code is [here](https://github.com/openshift/svt/tree/master/image_provisioner)

File: playbooks/image.yaml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+- name: Launch an instance to build images
+  hosts: localhost
+  connection: local
+  gather_facts: no
+  tasks:
+    - name: Create a persistent disk for the image
+      gce_pd:
+        service_account_email: "{{ gce_service_account }}"
+        credentials_file: "{{ playbook_dir + '/files/' + gcs_registry_keyfile }}"
```
Comment (stevekuznetsov) [inline playbooks/image.yaml:11]: Nit, but string concatenation was supposed to be `~` I thought in Jinja.

File: playbooks/image.yaml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        project_id: "{{ gce_project_id }}"
+        zone: "{{ gce_zone_name }}"
+        name: "{{ provision_prefix }}build-image-instance"
+        disk_type: pd-ssd
+        image: "{{ provision_gce_base_image | default('https://www.googleapis.com/compute/v1/projects/rhel-cloud/global/images/rhel-7-v20170110') }}"
+        size_gb: 10
```
Comment (stevekuznetsov) [inline playbooks/image.yaml:17]: This is probably the parameter we end up tweaking the most for image creation, next to RPM dependency list.

File: playbooks/image.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
+# This playbook ensures that a base image is up to date with all of the required settings
+
+- name: Launch an instance to build images
+  hosts: localhost
+  connection: local
+  gather_facts: no
+  tasks:
+    - name: Create a persistent disk for the image
+      gce_pd:
```
Comment (stevekuznetsov) [inline playbooks/image.yaml:9]: Are we not using different disks for `/root` and Docker storage?

**Review by stevekuznetsov — COMMENTED** [review]
I think in general the approach we want for image creation will be tiered. I'm not sure we really need much other than RHEL in the "real" base image... as:

We have needs for an AMI both with and without Docker, for instance Origin tests that need a Daemon socket like `test/end-to-end` -- when those tests run on Origin, we won't be able to mount in the host' socket, so I think we will need to provision from an image with Docker. Tests that need an API server like `test/cmd` could just start another pod or another container in the same pod, and don't need a VM. Tests that provision a cluster won't need Docker on their AMIs as the installer can take care of it. 

The current suite of quality of life RPM installs (`screen`, `tree`) may be nice for Origin extended tests running against a cluster and definitely are nice for devenvs, but the performance testing team doesn't seem to need them.

Go should be a non-contender except for the devenv.

Repositories again should not be needed except the devenv ... so based on my exposure to what we really need in a gold image if we want to be solving a broader problem the base image should just be RHEL. I'm not sure there is a huge merit to slapping a disk in there -- that should be under the purview of the provisioner logic, not the image minting logic, which in this repo are confounded.

**Comment by stevekuznetsov** [comment]
Regarding `yum update`.... maybe. In general I think the approach we have of doing things like this in our jobs sucks:

```
# Print useful version information
go version
sudo docker version
```

We should know what versions we have by looking at the tools that have assembled a job environment and run it ... and `yum update` only makes that harder to determine. I would really appreciate some broader understanding of how many times a `yum update` (as opposed to selectively bumping specific RPMs) has been useful for us.

**Comment by smarterclayton** [comment]
I do think we want to be testing and using the latest packages in images *except* for the ones we care about.  Agree that picking up major core dependencies suddenly is surprising, but one of the advantages is that it simplifies management and staying up to date.  So arguments both ways.

**Comment by smarterclayton** [comment]
> I think in general the approach we want for image creation will be tiered. I'm not sure we really need much other than RHEL in the "real" base image... as:

The image being built here is the "gold image" for whatever definition of gold their process uses - i.e. if they're doing autoscaling groups, the gold image needs a full install.  If they're doing a regular install, it just needs to be able to run the ansible tasks.  That choice - "what is this image for" seems like it's bimodal for this code.

> We have needs for an AMI both with and without Docker, for instance Origin tests that need a Daemon socket like test/end-to-end -- when those tests run on Origin, we won't be able to mount in the host' socket, so I think we will need to provision from an image with Docker. Tests that need an API server like test/cmd could just start another pod or another container in the same pod, and don't need a VM. Tests that provision a cluster won't need Docker on their AMIs as the installer can take care of it.

This is for a running cluster, so it needs the container manager.  Ansible can do that, but depending on the bimodal state the goal is either "everything" or "most everything".  It also saves time during install.

> The current suite of quality of life RPM installs (screen, tree) may be nice for Origin extended tests running against a cluster and definitely are nice for devenvs, but the performance testing team doesn't seem to need them.

Agree

> Go should be a non-contender except for the devenv.

Agree

> Repositories again should not be needed except the devenv ... so based on my exposure to what we really need in a gold image if we want to be solving a broader problem the base image should just be RHEL. I'm not sure there is a huge merit to slapping a disk in there -- that should be under the purview of the provisioner logic, not the image minting logic, which in this repo are confounded.

The disk in this example is because it's required to commit an image on GCE.

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Pull request merged**